src/storage/src/source/mod.rs

```diff
@@ -149,6 +149,28 @@ where
     pub headers: Option<Vec<(String, Option<Vec<u8>>)>>,
 }
 
+impl<D: Data, V: Data> From<SourceOutput<(), V>> for SourceOutput<Option<D>, V> {
```

Comment:
Hm, I'm not sure if this conversion is a concern of `SourceOutput`. This equivalence only makes sense when talking about SourceTypes, which is where the constraint flows from. What do you think of a wrapper type that describes this logic at the source type level like so?:

```rust
/// A wrapper that converts a delimieted source reader that only provides
/// values into a key/value reader whose key is always None
struct DelimitedValueSource<S>(S);

impl<S> SourceReader for DelimitedValueSource<S>
where
    S: SourceReader<Key = (), Value = Option<Vec<u8>>>,
{
    type Key = Option<Vec<u8>>;
    type Value = Option<Vec<u8>>;

    fn new(
        source_name: String,
        source_id: SourceInstanceId,
        worker_id: usize,
        worker_count: usize,
        consumer_activator: SyncActivator,
        connector: ExternalSourceConnector,
        aws_external_id: AwsExternalId,
        restored_offsets: Vec<(PartitionId, Option<MzOffset>)>,
        encoding: SourceDataEncoding,
        logger: Option<Logger>,
        metrics: crate::source::metrics::SourceBaseMetrics,
    ) -> Result<Self, anyhow::Error> {
        Self(S::new(todo!()))
    }

    fn get_next_message(&mut self) -> Result<NextMessage<Self::Key, Self::Value>, anyhow::Error> {
        match self.0.get_next_message()? {
            NextMessage::Ready(msg) => Ok(NextMessage::Ready(SourceMessage { key: None })),
            _ => todo!(),
        }
    }
}
```

Then when you instantiate the source above it will become:

```rust
source::create_source::<_, DelimitedValueSource<KinesisSourceReader>>(
```